### PR TITLE
Cookie extractor from Set-Cookie response.

### PR DESCRIPTION
**Pull request by danielatdattrixdotcom**
I needed this for working with an API that uses Django auth/sessions to work with CSRF cookie and X-CSRFToken header.

### CODE REVIEW

File: pyresttest/validators.py [hunk 1]
```diff
@@ -289,6 +290,24 @@ def parse(cls, config, extractor_base=None):
         return cls.configure_base(config, base)
 
 
+class CookieExtractor(AbstractExtractor):
+    """ Extractor that pulls out a named cookie """
+    extractor_type = 'cookie'
+    is_header_extractor = True
```
Comment (svanoort) [inline pyresttest/validators.py]: Excellent!  Glad you picked up and added this.

File: pyresttest/validators.py [hunk 2]
```diff
@@ -289,6 +290,24 @@ def parse(cls, config, extractor_base=None):
...
+
+    def extract_internal(self, query = None, args = None, body = None, headers = None):
+        extracted = [y[1] for y in filter(lambda x: x[0] == 'set-cookie', headers) if y[1].startswith(query + '=')]
+        if len(extracted) == 0:
+            raise ValueError("Cookie named {0} not known".format(query))
```
Comment (svanoort) [inline pyresttest/validators.py]: Good.

File: pyresttest/validators.py [hunk 3]
```diff
@@ -289,6 +290,24 @@ def parse(cls, config, extractor_base=None):
...
+        else:
+            return str(dict(Cookie.SimpleCookie(extracted[0]))[query].value)
```
Comment (svanoort) [inline pyresttest/validators.py]: Please, can you explain what is happening here?   It feels like some of the casting/here may not be needed.   Also, how will this handle if the same cookie is set multiple times?  (Which IIRC is legal in HTTP, if probably unlikely.)

Comment (svanoort) [inline pyresttest/validators.py]: Two other points:  

* This needs to be Python 3 safe, I think *probably* best not to cast to string and use the original type for the header.  I'm worried about potential for encoding mishaps here (let the initial processing deal with it). 
* I *think* it's best here to allow for returning a list of cookie values if set multiple times.

**Comment by svanoort** [comment]
@danielatdattrixdotcom Thank you, this is a feature that I know is in fairly high demand.  I've added a couple comments (small changes) before this is ready to be merged.  It also needs some sort of small unit test in test_validators.py.   

This weekend I'm going to play with it a bit more and look more closely at the format of output (it might make sense to add a few options).

You get kudos by the way for looking closely at how the other extractors are implemented and following that (and adding documentation as well)!

File: pyresttest/validators.py (hunk 3 above)
Comment (danielatdattrixdotcom) [inline pyresttest/validators.py]: Good point on the casting, it is not needed. Just an artefact for one of my iterations. Without it I think it will read straight-forward, just looping and looking for the header value starting with the case-sensitive name matching query.  So casting is about to go away.

As for the multiple headers. RFC 6265 states **"Servers SHOULD NOT include more than one Set-Cookie header field in the same response with the same cookie-name."**. And the suggested way for agents to handle this is only accept the last Set-Cookie header sent for a given name/path. I did not see how to pass args to an extractor in the config files, I see it done in the tests though. If you can give me the syntax on that I think a **path** arg to the extractor would be the best way to handle that situation instead of a list.

**Comment by danielatdattrixdotcom** [comment]
@svanoort Let me know on the args/options to extractors in the YAML/JSON files and your thoughts on that and I'll make the updates and tests.

**Comment by alswl** [comment]
Oh no, I had write the same feature code, but @danielatdattrixdotcom 's documents is nice.
😂

File: pyresttest/validators.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import logging
 import json
+import Cookie
```
Comment (svanoort) [inline pyresttest/validators.py]: Fair, but we need to do the following for python 3 compatibility, since the library changed in Py3 :cry:  https://docs.python.org/2/library/cookie.html

if 'sys' is not in imports, add it:

```python
PYTHON_MAJOR_VERSION = sys.version_info[0]
if PYTHON_MAJOR_VERSION > 2:
    import http.cookies as Cookie
else:
   import Cookie
```

File: pyresttest/validators.py [hunk 5]
```diff
@@ -289,6 +290,24 @@ def parse(cls, config, extractor_base=None):
...
+    """ Extractor that pulls out a named cookie """
+    extractor_type = 'cookie'
+    is_header_extractor = True
+
+    def extract_internal(self, query = None, args = None, body = None, headers = None):
+        extracted = [y[1] for y in filter(lambda x: x[0] == 'set-cookie', headers) if y[1].startswith(query + '=')]
+        if len(extracted) == 0:
+            raise ValueError("Cookie named {0} not known".format(query))
+        else:
+            return Cookie.SimpleCookie(extracted[-1])[query].value
```
Comment (svanoort) [inline pyresttest/validators.py]: I'm not as familiar with the syntax of the cookie library, but this *seems* solid.  Would have to test it a bit though.

**Comment by svanoort** [comment]
@danielatdattrixdotcom Coming back to this now that the crunch period at work is settling down (spoiler: Jenkins, and specifically work on Pipeline Stage View plugin):  I think it's not worth adding options to the extractor right now - I had a notion of eventually allowing use of regexes or other matchers in extraction, but it's not useful yet.  We need support for templating with collections (or transformations on collections), and/or some embedded transformation language, if we're going to go down that route.  It's planned in the future, but too far off to worry about now. 

Added one comment about a Python 3 compatibility challenge (just an import change).

So, from my point of view, to get to merge, what we need are automated tests, covering a couple key edge cases: cookie doesn't exist, cookie exists with multiple values, normal cookie case.  This means a unit test in test_validators.py, and a functional test in functionaltest.py -- the unit test just covers the extractor parsing and behavior (you can use some cookie info taken from a browser session).  

The functional test needs to work with real HTTP requests; I'd prefer to add cookie-setting to the test Django-tastypie app (probably easiest done with the sessions middleware option), but if that proves hard we can just invoke a public API or site, as long as the request is trivial.  I've been using GitHub for this in the past.

File: pyresttest/validators.py (hunk 5 above)
Comment (alswl) [inline pyresttest/validators.py]: it return the last index of cookies. The behaviors may be strange.

```
if len(extracted) != 0:
    raise ValueError('xxx')
return Cookie.SimpleCookie(extracted[0])[query].value 
```

Is this better? fast fail.

Comment (svanoort) [inline pyresttest/validators.py]: > it return the last index of cookies. The behaviors may be strange.

@alswl What do you mean by this?  The -1 array index is fine, and we *should* be using the last cookie specified, I'm just wondering about the Cookie.SimpleCookie there and what its failure modes are. 

I think you also meant this, which doesn't end up any different than the original in truth (it is slightly cleaner):

```python
if len(extracted) == 0:
    raise ValueError('xxx')
return Cookie.SimpleCookie(extracted[-1])[query].value 
```

File: pyresttest/validators.py (hunk 4 above)
Comment (jimi-c) [inline pyresttest/validators.py]: Even easier, just do this:
```from .six.moves import http_cookies as Cookie
```
I'm testing this patch with the above and it works perfectly well on both Py2 and Py3:
```
$ python2 $(which pyresttest) http://localhost:8000 test/api.yml 
Test Group Default SUCCEEDED: : 5/5 Tests Passed!

$ python3 $(which pyresttest) http://localhost:8000 test/api.yml 
Test Group Default SUCCEEDED: : 5/5 Tests Passed!
```
